[serializer] Add config param to use v2 api

olivielpeau opened the pull request:
### What does this PR do?

* Adds a config param to use v2 api, per endpoint
* Adds a `Content-Encoding` header to the payloads when they're compressed

### Motivation

This should allow us to start testing the v2 endpoints one by one. The default endpoints will remain the v1 endpoints until the v2 api is stable.

### Additional Notes

See commit descriptions.

Example config to use the v2 api on the `series` endpoint only:

```yaml
use_v2_endpoint:
  series: true
```

The options are not documented directly in `datadog.yaml` because I don't think customers should be changing them, they're there for our internal testing purposes only (at least for now).

**[update]**

I've disabled compression on the v1 endpoints as we've noticed the agent6 zstd compression doesn't work on the v1 endpoints right now (`400` status code). I've added a card to our backlog to track this, we can spend more time investigating this as part of a separate PR.

olivielpeau committed:
[serializer] Add config option to switch to v2 endpoint

Per endpoint config option.

For now we use the v1 endpoints by default, until the v2 endpoints are
considered stable.

olivielpeau committed:
[serializer] Add `Content-Encoding` header when compression is used

The intake uses that header to determine if/how to decompress the
payloads, so let's add it in the serializer.

olivielpeau committed:
[serializer] Rename `endpoint` -> `api`

gmmeyer reviewed (approved):
I have a few questions but generally think this is good!

gmmeyer reviewed (changes requested):
Left two more comments about this

olivielpeau committed:
[serializer] Address review comments

olivielpeau committed:
[serializer] Disable compression on v1 series/check_run endpoints

zstd compression is supposed to work on the v1 series and check_run
endpoints, but for some reason it doesn't. Let's disable it until we
have time to investigate more.

gmmeyer reviewed (approved):
🎉
